### PR TITLE
Research

### PR DESCRIPTION
**Pull request by AlexeyKozhevin**
- Remove `framework` parameter
- Separation of `tf_device`/`torch_device` keys in config
- Add profiling notebooks
- Remove folder names iteration

### CODE REVIEW

File: batchflow/research/workers.py [hunk 1]
```diff
@@ -12,17 +12,21 @@ def init(self):
...
-        elif self.framework == 'torch':
-            prefix = 'cuda:'
-        else:
-            raise ValueError('Unknown framework: {}'.format(self.framework))
+        tf_prefix = '/device:GPU:'
+        torch_prefix = 'cuda:'
 
         if len(self.gpu) <= 1:
-            self.gpu_configs = [dict(device=prefix+'0') for i in range(n_branches)]
+            self.gpu_configs = [dict(
```
Comment (roman-kh) [inline batchflow/research/workers.py]: Update models with unified device naming

File: examples/experiments/profiling/torch_research.py [hunk 2]
```diff
@@ -44,7 +44,7 @@
                            fetches='predictions',
                            save_to=V('predictions'))
             .gather_metrics('class', targets=B('labels'), predictions=V('predictions'),
-                                fmt='logits', axis=-1, save_to=V('metrics'), mode='a')
+                                fmt='logits', axis=-1, save_to=V('metrics', mode='a'))
```
Comment (roman-kh) [inline examples/experiments/profiling/torch_research.py]: This might contain excessive spaces

File: examples/experiments/profiling/tf_research.py [hunk 3]
```diff
@@ -47,7 +47,7 @@
                          feed_dict={'images': B('images'), 'labels': B('labels')},
                          save_to=V('predictions'))
             .gather_metrics('class', targets=B('labels'), predictions=V('predictions'),
-                                fmt='logits', axis=-1, save_to=V('metrics'), mode='a')
+                                fmt='logits', axis=-1, save_to=V('metrics', mode='a'))
```
Comment (roman-kh) [inline examples/experiments/profiling/tf_research.py]: Aren't there too many spaces?

File: examples/experiments/profiling/tf_model.py [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+
+test_ppl = (mnist.test.p
+    .init_variable('predictions') 
+    .init_variable('metrics', init_on_each_run=None) 
+    .import_model('conv', train_ppl)
+    .to_array()
+    .predict_model('conv', 
+                   fetches='predictions', 
+                   feed_dict={'images': B('images'), 'labels': B('labels')},
```
Comment (roman-kh) [inline examples/experiments/profiling/tf_model.py]:
```suggestion
                   images=B('images'), labels=B('labels'),
```

File: examples/experiments/profiling/tf_model.py [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+mnist = MNIST()
+
+train_ppl = (mnist.train.p
+    .init_variable('loss', init_on_each_run=list)
+    .init_variable('accuracy', init_on_each_run=list)
+    .init_model('dynamic', VGG16, 'conv', config=model_config)
+    .to_array()
+    .train_model('conv', 
+                 fetches='loss', 
+                 feed_dict={'images': B('images'), 'labels': B('labels')},
```
Comment (roman-kh) [inline examples/experiments/profiling/tf_model.py]:
```suggestion
                   images=B('images'), labels=B('labels'),
```

File: examples/experiments/profiling/tf_model.py (hunk 4 above)
Comment (roman-kh) [reply examples/experiments/profiling/tf_model.py]: Do you really need labels here?

File: examples/experiments/profiling/tf_research.py [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+test_root = mnist.test.p.run(BATCH_SIZE, shuffle=True, n_epochs=1, lazy=True)
+
+train_template = (Pipeline()
+    .init_variable('loss', init_on_each_run=list)
+    .init_variable('accuracy', init_on_each_run=list)
+    .init_model('dynamic', VGG16, 'conv', config=model_config)
+    .to_array()
+    .train_model('conv',
+                fetches='loss',
+                feed_dict={'images': B('images'), 'labels': B('labels')},
```
Comment (roman-kh) [inline examples/experiments/profiling/tf_research.py]:
```suggestion
                   images=B('images'), labels=B('labels'),
```

File: examples/experiments/profiling/tf_research.py [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+)
+
+test_template = (Pipeline()
+    .init_variable('predictions')
+    .init_variable('metrics', init_on_each_run=None)
+    .import_model('conv', C('import_from'))
+    .to_array()
+    .predict_model('conv',
+                    fetches='predictions',
+                    feed_dict={'images': B('images'), 'labels': B('labels')},
```
Comment (roman-kh) [inline examples/experiments/profiling/tf_research.py]:
```suggestion
                   images=B('images'), labels=B('labels'),
```

Comment (roman-kh) [reply examples/experiments/profiling/tf_research.py]: Do you really need labels here?

File: examples/experiments/profiling/tf_model.py [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+from batchflow.opensets import MNIST
+from batchflow.models.tf import VGG16
+from batchflow.research import Research, Option
+
+BATCH_SIZE=64
+
+model_config={
+    'session/config': tf.ConfigProto(allow_soft_placement=True),
+    'inputs/images/shape': (28, 28, 1),
+    'inputs/labels': {
```
Comment (roman-kh) [inline examples/experiments/profiling/tf_model.py]:
```suggestion
    'inputs/labels/classes': D('num_classes'),
```

File: examples/experiments/profiling/tf_model.py [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+        'classes': D('num_classes'),
+        'transform': 'ohe',
```
Comment (roman-kh) [inline examples/experiments/profiling/tf_model.py]: This is not needed at all

File: examples/experiments/profiling/tf_model.py [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+        'name': 'targets'
```
Comment (roman-kh) [inline examples/experiments/profiling/tf_model.py]: This is not needed at all

File: examples/experiments/profiling/tf_model.py [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+sys.path.append("../../..")
+from batchflow import Pipeline, B, C, V, D
+from batchflow.opensets import MNIST
+from batchflow.models.tf import VGG16
+from batchflow.research import Research, Option
+
+BATCH_SIZE=64
+
+model_config={
+    'session/config': tf.ConfigProto(allow_soft_placement=True),
```
Comment (roman-kh) [inline examples/experiments/profiling/tf_model.py]: Remove as it will be obsolete soon

File: examples/experiments/profiling/tf_model.py [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+    .run(BATCH_SIZE, shuffle=True, n_epochs=1, lazy=True))
+
+
+test_ppl = (mnist.test.p
+    .init_variable('predictions') 
+    .init_variable('metrics', init_on_each_run=None) 
+    .import_model('conv', train_ppl)
+    .to_array()
+    .predict_model('conv',
+                   images=B('images'), labels=B('labels'),
```
Comment (roman-kh) [inline examples/experiments/profiling/tf_model.py]:
```suggestion
                   images=B('images'),
```
labels aren't used anywhere

File: examples/experiments/profiling/tf_research.py [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+sys.path.append("../../..")
+from batchflow import Pipeline, B, C, V
+from batchflow.opensets import MNIST
+from batchflow.models.tf import VGG16
+from batchflow.research import Research
+
+BATCH_SIZE = 64
+
+model_config = {
+    'session/config': tf.ConfigProto(allow_soft_placement=True),
```
Comment (roman-kh) [inline examples/experiments/profiling/tf_research.py]: Remove

File: examples/experiments/profiling/tf_research.py [hunk 14]
```diff
@@ -0,0 +1,70 @@
...
+    'inputs/images/shape': (28, 28, 1),
+    'inputs/labels': {
```
Comment (roman-kh) [inline examples/experiments/profiling/tf_research.py]: D('num_classes')

File: examples/experiments/profiling/tf_research.py [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+                save_to=V('loss', mode='w'))
+)
+
+test_template = (Pipeline()
+    .init_variable('predictions')
+    .init_variable('metrics', init_on_each_run=None)
+    .import_model('conv', C('import_from'))
+    .to_array()
+    .predict_model('conv',
+                    images=B('images'), labels=B('labels'),
```
Comment (roman-kh) [inline examples/experiments/profiling/tf_research.py]: labels

File: examples/experiments/profiling/torch_model.py [hunk 16]
```diff
@@ -0,0 +1,48 @@
...
+sys.path.append("../../..")
+from batchflow import B, V
+from batchflow.opensets import MNIST
+from batchflow.models.torch import VGG16
+
+BATCH_SIZE = 64
+
+model_config = {
+    'inputs/images/shape': (1, 28, 28),
+    'inputs/labels': {
```
Comment (roman-kh) [inline examples/experiments/profiling/torch_model.py]: D('num_classes')

File: examples/experiments/profiling/torch_model.py [hunk 17]
```diff
@@ -0,0 +1,48 @@
...
+                 save_to=V('loss', mode='w'))
+    .run(BATCH_SIZE, shuffle=True, n_epochs=1, lazy=True))
+
+
+test_ppl = (mnist.test.p
+    .init_variable('predictions')
+    .init_variable('metrics', init_on_each_run=None)
+    .import_model('conv', train_ppl)
+    .to_array(channels='first', dtype='float32')
+    .predict_model('conv', B('images'), targets=B('labels'),
```
Comment (roman-kh) [inline examples/experiments/profiling/torch_model.py]: targets are not necessary here

File: examples/experiments/profiling/torch_research.py [hunk 18]
```diff
@@ -0,0 +1,66 @@
...
+from batchflow import Pipeline, B, C, V
+from batchflow.opensets import MNIST
+from batchflow.models.torch import VGG16
+from batchflow.research import Research, Option
+
+BATCH_SIZE = 64
+
+model_config = {
+    'inputs/images/shape': (1, 28, 28),
+    'inputs/labels': {
```
Comment (roman-kh) [inline examples/experiments/profiling/torch_research.py]: D('num_classes')

File: examples/experiments/profiling/torch_research.py [hunk 19]
```diff
@@ -0,0 +1,66 @@
...
+    .train_model('conv', B('images'), B('labels'),
+                fetches='loss', save_to=V('loss', mode='w'))
+)
+
+test_template = (Pipeline()
+    .init_variable('predictions')
+    .init_variable('metrics', init_on_each_run=None)
+    .import_model('conv', C('import_from'))
+    .to_array(channels='first', dtype='float32')
+    .predict_model('conv', B('images'), targets=B('labels'),
```
Comment (roman-kh) [inline examples/experiments/profiling/torch_research.py]:
```suggestion
    .predict_model('conv', B('images'),
```

File: examples/experiments/profiling/torch_research.py [hunk 20]
```diff
@@ -0,0 +1,66 @@
...
+                   fetches='predictions',
+                   save_to=V('predictions'))
+    .gather_metrics('class', targets=B('labels'), predictions=V('predictions'),
+                    fmt='logits', axis=-1, save_to=V('metrics', mode='a'))
+)
+
+train_ppl = train_root + train_template
+test_ppl = test_root + test_template
+
+def get_accuracy(iteration, experiment, pipeline):
```
Comment (roman-kh) [inline examples/experiments/profiling/torch_research.py]: Let's add it as a method of `Research`: 
research
.add_pipeline(...)
.get_metrics(pipeline_name, metrics_var, metrics_name, execute,...)

File: examples/tutorials/10_cross_validation.ipynb [hunk 21]
```diff
@@ -107,7 +106,7 @@
     "                   labels={'classes': 10, 'transform': 'ohe', 'name': 'targets'}),\n",
```
Comment (roman-kh) [inline examples/tutorials/10_cross_validation.ipynb]: D('num_classes')

**Comment by roman-kh** [comment]
Torch and TF models seem to have different device namings again.
